# Running EA pipeline

## EAinit

```{bash, eval=FALSE}
EAinit Path/to/a/DNAnexus/result/folder
```

The execution of above command will create a sub-folder (QuickOmics_[timestamp]) in the specified DNAnexus result folder.
There will be five files in the result folder:

- compareInfo.csv: an empty comparison definition file (with header). Please fill in this file before ```EArun``` call.
- config.yml: a config fill specifies the parameters of the ```EAqc``` and ```EArun```. Please update **covariates_adjust** after ```EAqc```.
- geneAnnotation.csv: a gene annotation file including gene symbol.
- sampleMeta.csv: a sample meta information file, please feel free to add additional columns whose column names should be considered to be added into **covariates_check** in *config.yml*.
- alignQC.pdf: plots generated from alignment QC metrics.

**_Please pay attention on the std out messages._**

## EAqc

```
EAqc Path/to/a/config/file
```
The execution of the command with the above default config file, expression PC analysis will be done against covariates specified in **covariates_check** in *config.yml* file. An excel file will list p-value for all numeric and categorical covariates, and with significant ones will be in plot pdf files. The analysis before covariate adjusting will have prefix *covariatePCanalysis_noAdjust*. 
Based on the above results, you can add covariates into **covariates_adjust** in *config.yml* file, and run ```EAqc``` again. This time additional expression PC analysis will be applied to covariate adjusted expression with files started with *covariatePCanalysis_Adjusted*. 

**_Please pay attention on the std out messages._**

## EArun

```
EArun Path/to/a/config/file
```
**_Please fill the compareInfo.csv before executing the above command_**

The execution of above command will produce R object for QuickOmics webserver to load. The process will generate the covariate adjusted logTPM for visualization; complete differentially expressed gene analysis and gene network generation. 

The results (three files) currently will need to be copied to a folder on ngs, in order to have web accession. **_Please pay attention on the std out messages._**