<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 7 Advanced plotting functions | RNASequest: An end-to-end reproducible RNAseq data analysis and publishing framework</title>
  <meta name="description" content="Chapter 7 Advanced plotting functions | RNASequest: An end-to-end reproducible RNAseq data analysis and publishing framework" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 7 Advanced plotting functions | RNASequest: An end-to-end reproducible RNAseq data analysis and publishing framework" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 7 Advanced plotting functions | RNASequest: An end-to-end reproducible RNAseq data analysis and publishing framework" />
  
  
  



<meta name="date" content="2024-03-28" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="output.html"/>
<link rel="next" href="quickomics-component.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">RNASequest</a></li>

<li class="divider"></li>
<li><a href="index.html#preface" id="toc-preface"><span class="toc-section-number">1</span> Preface</a></li>
<li><a href="introduction.html#introduction" id="toc-introduction"><span class="toc-section-number">2</span> Introduction</a></li>
<li><a href="installation.html#installation" id="toc-installation"><span class="toc-section-number">3</span> Installation</a>
<ul>
<li><a href="installation.html#install-expressionanalysis" id="toc-install-expressionanalysis"><span class="toc-section-number">3.1</span> Install ExpressionAnalysis</a></li>
<li><a href="installation.html#install-quickomics" id="toc-install-quickomics"><span class="toc-section-number">3.2</span> Install Quickomics</a>
<ul>
<li><a href="installation.html#install-r-packages" id="toc-install-r-packages"><span class="toc-section-number">3.2.1</span> Install R packages</a></li>
<li><a href="installation.html#clone-quickomics-github-repository" id="toc-clone-quickomics-github-repository"><span class="toc-section-number">3.2.2</span> Clone Quickomics GitHub repository</a></li>
<li><a href="installation.html#launch-the-r-shiny-app" id="toc-launch-the-r-shiny-app"><span class="toc-section-number">3.2.3</span> Launch the R Shiny App</a></li>
<li><a href="installation.html#configure-quickomics-to-load-projects-directly-from-urls" id="toc-configure-quickomics-to-load-projects-directly-from-urls"><span class="toc-section-number">3.2.4</span> Configure QuickOmics to load projects directly from URLs</a></li>
<li><a href="installation.html#prepare-project-data-to-be-loaded-from-url" id="toc-prepare-project-data-to-be-loaded-from-url"><span class="toc-section-number">3.2.5</span> Prepare project data to be loaded from URL</a></li>
</ul></li>
</ul></li>
<li><a href="data-preparation.html#data-preparation" id="toc-data-preparation"><span class="toc-section-number">4</span> Data preparation</a>
<ul>
<li><a href="data-preparation.html#retrieving-data-from-ncbi-geo" id="toc-retrieving-data-from-ncbi-geo"><span class="toc-section-number">4.1</span> Retrieving data from NCBI GEO</a></li>
<li><a href="data-preparation.html#file-hierarchy" id="toc-file-hierarchy"><span class="toc-section-number">4.2</span> File hierarchy</a></li>
<li><a href="data-preparation.html#self-prepared-files" id="toc-self-prepared-files"><span class="toc-section-number">4.3</span> Self-prepared files</a></li>
</ul></li>
<li><a href="expression-analysis-ea-component.html#expression-analysis-ea-component" id="toc-expression-analysis-ea-component"><span class="toc-section-number">5</span> Expression analysis (EA) component</a>
<ul>
<li><a href="expression-analysis-ea-component.html#eainit" id="toc-eainit"><span class="toc-section-number">5.1</span> EAinit</a></li>
<li><a href="expression-analysis-ea-component.html#eaqc" id="toc-eaqc"><span class="toc-section-number">5.2</span> EAqc</a></li>
<li><a href="expression-analysis-ea-component.html#earun" id="toc-earun"><span class="toc-section-number">5.3</span> EArun</a></li>
<li><a href="expression-analysis-ea-component.html#eareport" id="toc-eareport"><span class="toc-section-number">5.4</span> EAreport</a></li>
<li><a href="expression-analysis-ea-component.html#ea2da" id="toc-ea2da"><span class="toc-section-number">5.5</span> EA2DA</a></li>
<li><a href="expression-analysis-ea-component.html#administration" id="toc-administration"><span class="toc-section-number">5.6</span> Administration</a></li>
</ul></li>
<li><a href="output.html#output" id="toc-output"><span class="toc-section-number">6</span> Output</a>
<ul>
<li><a href="output.html#quickomics-visualization" id="toc-quickomics-visualization"><span class="toc-section-number">6.1</span> QuickOmics visualization</a></li>
<li><a href="output.html#other-output-files" id="toc-other-output-files"><span class="toc-section-number">6.2</span> Other output files</a></li>
</ul></li>
<li><a href="advanced-plotting-functions.html#advanced-plotting-functions" id="toc-advanced-plotting-functions"><span class="toc-section-number">7</span> Advanced plotting functions</a>
<ul>
<li><a href="advanced-plotting-functions.html#pca-plot" id="toc-pca-plot"><span class="toc-section-number">7.1</span> PCA plot</a></li>
<li><a href="advanced-plotting-functions.html#covariate-checking" id="toc-covariate-checking"><span class="toc-section-number">7.2</span> Covariate checking</a></li>
<li><a href="advanced-plotting-functions.html#volcano-plot" id="toc-volcano-plot"><span class="toc-section-number">7.3</span> Volcano plot</a></li>
<li><a href="advanced-plotting-functions.html#heatmap" id="toc-heatmap"><span class="toc-section-number">7.4</span> Heatmap</a></li>
</ul></li>
<li><a href="quickomics-component.html#quickomics-component" id="toc-quickomics-component"><span class="toc-section-number">8</span> Quickomics component</a></li>
<li><a href="bookdown-component.html#bookdown-component" id="toc-bookdown-component"><span class="toc-section-number">9</span> Bookdown component</a></li>
<li><a href="slide-deck-component.html#slide-deck-component" id="toc-slide-deck-component"><span class="toc-section-number">10</span> Slide deck component</a></li>
<li><a href="shinyone-component.html#shinyone-component" id="toc-shinyone-component"><span class="toc-section-number">11</span> ShinyOne component</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">RNASequest: An end-to-end reproducible RNAseq data analysis and publishing framework</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="advanced-plotting-functions" class="section level1" number="7">
<h1><span class="header-section-number">Chapter 7</span> Advanced plotting functions</h1>
<p>For advanced users, EA offers versatile plotting functions to adjust your figures. Each plotting function within <a href="https://github.com/interactivereport/RNASequest/blob/main/src/Plot_Functions.R" target="_blank">Plot_Functions.R</a> has many options, mirroring the user interface from QuickOmics. Here are several examples:</p>
<div id="pca-plot" class="section level2" number="7.1">
<h2><span class="header-section-number">7.1</span> PCA plot</h2>
<pre><code># Load RData created from EArun 
# Here is an example
# https://github.com/interactivereport/Quickomics/blob/master/data/Mouse_microglia_RNA-Seq.RData 

load(&quot;data/Mouse_microglia_RNA-Seq.RData&quot;)

# Load Plot_Functions.R
# https://github.com/interactivereport/RNASequest/tree/main/src
setwd(&quot;~/RNASequest/src/&quot;)
source(&quot;Plot_Functions.R&quot;) 

# Create a basic PCA Plot. 
p&lt;-PCA_Plot(data_wide, MetaData) 
print(p)  </code></pre>
<p><a href="https://interactivereport.github.io/RNASequest/tutorial/images/PCA.png"><img src="https://interactivereport.github.io/RNASequest/tutorial/images/PCA.png" /></a></p>
<pre><code># PCA plot with additional options:
p&lt;-PCA_Plot(data_wide, MetaData, pcnum=c(2, 3),label_samples=F, 
            PCAcolorby=&quot;Age&quot;,  PCAshapeby=&quot;Genotype&quot;, ellipsoid=T, rug = T) 

# Plot only selected samples:
input_samples=MetaData$sampleid[20:60] 
p&lt;-PCA_Plot(data_wide, MetaData, input_samples) 
print(p)  </code></pre>
</div>
<div id="covariate-checking" class="section level2" number="7.2">
<h2><span class="header-section-number">7.2</span> Covariate checking</h2>
<pre><code># Check covariates:
# This can be displayed as a table in R Markdown

res&lt;-PC_covariates_out(data_wide, MetaData) 
sig_covariates=res$selVar_All                 #data frame to print as table</code></pre>
<p><a href="https://interactivereport.github.io/RNASequest/tutorial/images/Covariates.png"><img src="https://interactivereport.github.io/RNASequest/tutorial/images/Covariates.png" /></a></p>
<pre><code># Check covariates with more options if needed:

res&lt;-PC_covariates_out(data_wide, MetaData, input_samples, select_covariates=c(&quot;Age&quot;, &quot;Genotype&quot;, &quot;Gender&quot;),  PC_cutoff=3, FDR_cutoff=0.2)

(sig_covariates=res$selVar_All) </code></pre>
</div>
<div id="volcano-plot" class="section level2" number="7.3">
<h2><span class="header-section-number">7.3</span> Volcano plot</h2>
<pre><code># Loop through the comparisons. 

p&lt;-Volcano_Plot(results_long, ProteinGeneName, comparison=&quot;2yr_KOvs2yr_WT&quot;) 
print(p) </code></pre>
<p><a href="https://interactivereport.github.io/RNASequest/tutorial/images/Volcano.png"><img src="https://interactivereport.github.io/RNASequest/tutorial/images/Volcano.png" /></a></p>
<pre><code># More controls on volcano plot:

p&lt;-Volcano_Plot(results_long, ProteinGeneName, comparison=&quot;1yr_KOvs1yr_WT&quot;, Ngenes=40, rasterize=&quot;No&quot;, vlegendpos=&quot;right&quot;,  Max_Pvalue=30, Max_logFC=1.5) </code></pre>
</div>
<div id="heatmap" class="section level2" number="7.4">
<h2><span class="header-section-number">7.4</span> Heatmap</h2>
<pre><code># Heatmap  

# Overview from 1000 most variable genes 

p&lt;-Heatmap_Plot(data_wide, MetaData, results_long, ProteinGeneName,heatmap_subset=&quot;variable&quot;, maxgenes=1000) 

draw(p, merge_legend=T,  auto_adjust = FALSE)  </code></pre>
<p><a href="https://interactivereport.github.io/RNASequest/tutorial/images/Heatmap1.png"><img src="https://interactivereport.github.io/RNASequest/tutorial/images/Heatmap1.png" /></a></p>
<pre><code># Heatmap from DEGs of a comparison:
# This can be done for all comparisons. 

p&lt;-Heatmap_Plot(data_wide, MetaData, results_long, ProteinGeneName,heatmap_subset=&quot;subset&quot;, comparison=&quot;2yr_WTvs2mo_WT&quot;) 

draw(p, merge_legend=T,  auto_adjust = FALSE)  </code></pre>
<p><a href="https://interactivereport.github.io/RNASequest/tutorial/images/Heatmap2.png"><img src="https://interactivereport.github.io/RNASequest/tutorial/images/Heatmap2.png" /></a></p>
<pre><code># With more options to customize the heatmap 

p&lt;-Heatmap_Plot(data_wide, MetaData, results_long, ProteinGeneName,heatmap_subset=&quot;subset&quot;, comparison=&quot;2yr_WTvs2mo_WT&quot;,  heatmap_fccut=4, heatmap_pvalcut=0.01, heatmap_psel = &quot;Padj&quot;, heatmap_row_dend=FALSE, heatmap_annot=c(&quot;Age&quot;, &quot;Genotype&quot;) ) 

draw(p, merge_legend=T,  auto_adjust = FALSE) </code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="output.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="quickomics-component.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/interactivereport/RNASequest/edit/gh-pages/07-AdvancedPlottingFeatures.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/interactivereport/RNASequest/blob/gh-pages/07-AdvancedPlottingFeatures.Rmd",
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
