<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 Expression analysis (EA) component | RNASequest: An end-to-end reproducible RNAseq data analysis and publishing framework</title>
  <meta name="description" content="Chapter 5 Expression analysis (EA) component | RNASequest: An end-to-end reproducible RNAseq data analysis and publishing framework" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 Expression analysis (EA) component | RNASequest: An end-to-end reproducible RNAseq data analysis and publishing framework" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 Expression analysis (EA) component | RNASequest: An end-to-end reproducible RNAseq data analysis and publishing framework" />
  
  
  



<meta name="date" content="2024-03-28" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="data-preparation.html"/>
<link rel="next" href="output.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">RNASequest</a></li>

<li class="divider"></li>
<li><a href="index.html#preface" id="toc-preface"><span class="toc-section-number">1</span> Preface</a></li>
<li><a href="introduction.html#introduction" id="toc-introduction"><span class="toc-section-number">2</span> Introduction</a></li>
<li><a href="installation.html#installation" id="toc-installation"><span class="toc-section-number">3</span> Installation</a>
<ul>
<li><a href="installation.html#install-expressionanalysis" id="toc-install-expressionanalysis"><span class="toc-section-number">3.1</span> Install ExpressionAnalysis</a></li>
<li><a href="installation.html#install-quickomics" id="toc-install-quickomics"><span class="toc-section-number">3.2</span> Install Quickomics</a>
<ul>
<li><a href="installation.html#install-r-packages" id="toc-install-r-packages"><span class="toc-section-number">3.2.1</span> Install R packages</a></li>
<li><a href="installation.html#clone-quickomics-github-repository" id="toc-clone-quickomics-github-repository"><span class="toc-section-number">3.2.2</span> Clone Quickomics GitHub repository</a></li>
<li><a href="installation.html#launch-the-r-shiny-app" id="toc-launch-the-r-shiny-app"><span class="toc-section-number">3.2.3</span> Launch the R Shiny App</a></li>
<li><a href="installation.html#configure-quickomics-to-load-projects-directly-from-urls" id="toc-configure-quickomics-to-load-projects-directly-from-urls"><span class="toc-section-number">3.2.4</span> Configure QuickOmics to load projects directly from URLs</a></li>
<li><a href="installation.html#prepare-project-data-to-be-loaded-from-url" id="toc-prepare-project-data-to-be-loaded-from-url"><span class="toc-section-number">3.2.5</span> Prepare project data to be loaded from URL</a></li>
</ul></li>
</ul></li>
<li><a href="data-preparation.html#data-preparation" id="toc-data-preparation"><span class="toc-section-number">4</span> Data preparation</a>
<ul>
<li><a href="data-preparation.html#retrieving-data-from-ncbi-geo" id="toc-retrieving-data-from-ncbi-geo"><span class="toc-section-number">4.1</span> Retrieving data from NCBI GEO</a></li>
<li><a href="data-preparation.html#file-hierarchy" id="toc-file-hierarchy"><span class="toc-section-number">4.2</span> File hierarchy</a></li>
<li><a href="data-preparation.html#self-prepared-files" id="toc-self-prepared-files"><span class="toc-section-number">4.3</span> Self-prepared files</a></li>
</ul></li>
<li><a href="expression-analysis-ea-component.html#expression-analysis-ea-component" id="toc-expression-analysis-ea-component"><span class="toc-section-number">5</span> Expression analysis (EA) component</a>
<ul>
<li><a href="expression-analysis-ea-component.html#eainit" id="toc-eainit"><span class="toc-section-number">5.1</span> EAinit</a></li>
<li><a href="expression-analysis-ea-component.html#eaqc" id="toc-eaqc"><span class="toc-section-number">5.2</span> EAqc</a></li>
<li><a href="expression-analysis-ea-component.html#earun" id="toc-earun"><span class="toc-section-number">5.3</span> EArun</a></li>
<li><a href="expression-analysis-ea-component.html#eareport" id="toc-eareport"><span class="toc-section-number">5.4</span> EAreport</a></li>
<li><a href="expression-analysis-ea-component.html#ea2da" id="toc-ea2da"><span class="toc-section-number">5.5</span> EA2DA</a></li>
<li><a href="expression-analysis-ea-component.html#administration" id="toc-administration"><span class="toc-section-number">5.6</span> Administration</a></li>
</ul></li>
<li><a href="output.html#output" id="toc-output"><span class="toc-section-number">6</span> Output</a>
<ul>
<li><a href="output.html#quickomics-visualization" id="toc-quickomics-visualization"><span class="toc-section-number">6.1</span> QuickOmics visualization</a></li>
<li><a href="output.html#other-output-files" id="toc-other-output-files"><span class="toc-section-number">6.2</span> Other output files</a></li>
</ul></li>
<li><a href="advanced-plotting-functions.html#advanced-plotting-functions" id="toc-advanced-plotting-functions"><span class="toc-section-number">7</span> Advanced plotting functions</a>
<ul>
<li><a href="advanced-plotting-functions.html#pca-plot" id="toc-pca-plot"><span class="toc-section-number">7.1</span> PCA plot</a></li>
<li><a href="advanced-plotting-functions.html#covariate-checking" id="toc-covariate-checking"><span class="toc-section-number">7.2</span> Covariate checking</a></li>
<li><a href="advanced-plotting-functions.html#volcano-plot" id="toc-volcano-plot"><span class="toc-section-number">7.3</span> Volcano plot</a></li>
<li><a href="advanced-plotting-functions.html#heatmap" id="toc-heatmap"><span class="toc-section-number">7.4</span> Heatmap</a></li>
</ul></li>
<li><a href="quickomics-component.html#quickomics-component" id="toc-quickomics-component"><span class="toc-section-number">8</span> Quickomics component</a></li>
<li><a href="bookdown-component.html#bookdown-component" id="toc-bookdown-component"><span class="toc-section-number">9</span> Bookdown component</a></li>
<li><a href="slide-deck-component.html#slide-deck-component" id="toc-slide-deck-component"><span class="toc-section-number">10</span> Slide deck component</a></li>
<li><a href="shinyone-component.html#shinyone-component" id="toc-shinyone-component"><span class="toc-section-number">11</span> ShinyOne component</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">RNASequest: An end-to-end reproducible RNAseq data analysis and publishing framework</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="expression-analysis-ea-component" class="section level1" number="5">
<h1><span class="header-section-number">Chapter 5</span> Expression analysis (EA) component</h1>
<p>First, make sure the .env has been created under the src folder:</p>
<pre><code>ls ~/RNASequest/src/.env</code></pre>
<div id="eainit" class="section level2" number="5.1">
<h2><span class="header-section-number">5.1</span> EAinit</h2>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb17-1"><a href="expression-analysis-ea-component.html#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="ex">EAinit</span> Path/to/a/DNAnexus/result/folder</span>
<span id="cb17-2"><a href="expression-analysis-ea-component.html#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="expression-analysis-ea-component.html#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Example:</span></span>
<span id="cb17-4"><a href="expression-analysis-ea-component.html#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="ex">EAinit</span> ~/RNASequest/example/SRP199678</span></code></pre></div>
<p>Execution of the above command will create a sub-folder (EA[timestamp]) in the specified RNAseq result folder.
There will be five files in the result folder:</p>
<ul>
<li>compareInfo.csv: an empty comparison definition file (with header). Please fill in this file before the <code>EArun</code> call.</li>
<li>config.yml: a config file specifies the parameters of the <code>EAqc</code> and <code>EArun</code>. Please update <strong>covariates_adjust</strong> after <code>EAqc</code>.</li>
<li>geneAnnotation.csv: a gene annotation file including gene symbol.</li>
<li>sampleMeta.csv: a sample meta-information file, please feel free to add additional columns whose column names should be considered to be added into <strong>covariates_check</strong> in <em>config.yml</em>.</li>
<li>alignQC.pdf: plots generated from alignment QC metrics.</li>
</ul>
<p><strong><em>Please pay attention to the std out messages.</em></strong></p>
</div>
<div id="eaqc" class="section level2" number="5.2">
<h2><span class="header-section-number">5.2</span> EAqc</h2>
<pre><code>EAqc Path/to/a/config/file

#Example:
EAqc ~/RNASequest/example/SRP199678/EA20220328_0/config.yml</code></pre>
<p>Through executing the command with the above default config file, expression PC analysis will be done against covariates specified in <strong>covariates_check</strong> in the <em>config.yml</em> file. An Excel file will list p-values for all numeric and categorical covariates, and significant ones will be in plot pdf files. The analysis before covariate adjusting will have the prefix <em>covariatePCanalysis_noAdjust</em>.
Based on the above results, you can add covariates into <strong>covariates_adjust</strong> in the <em>config.yml</em> file, and rerun <code>EAqc</code>. This time additional expression PC analysis will be applied to covariate-adjusted expression with files started with <em>covariatePCanalysis_Adjusted</em>.</p>
<p><strong><em>Please pay attention to the std out messages.</em></strong></p>
</div>
<div id="earun" class="section level2" number="5.3">
<h2><span class="header-section-number">5.3</span> EArun</h2>
<pre><code>EArun Path/to/a/config/file

#Example:
EArun ~/RNASequest/example/SRP199678/EA20220328_0/config.yml</code></pre>
<p><strong><em>Please fill in the compareInfo.csv before executing the above command</em></strong></p>
<p>Execution of the above command will produce R objects for QuickOmics webserver to load. The process will generate the covariate-adjusted logTPM for visualization; complete differentially expressed gene analysis and gene network generation.</p>
<p>The results (four files) can be uploaded to the QuickOmics webserver.</p>
<p><strong><em>Please pay attention to the std out messages.</em></strong></p>
</div>
<div id="eareport" class="section level2" number="5.4">
<h2><span class="header-section-number">5.4</span> EAreport</h2>
<pre><code>EAreport Path/to/a/config/file

#Example:
EAreport ~/RNASequest/example/SRP199678/EA20220328_0/config.yml</code></pre>
<p>By running the command above, the pipeline will generate a <strong>BookdownReport</strong> folder in the same directory as the config file. This folder contains the raw Rmd files, as well as the final bookdown report, which is the <strong>BookdownReport/docs/index.html</strong> file. If you would like to send the full report to your collaborators, please download the tarball created under the EA working directory, named as <strong>ProjectName_BookdownReport.tar.gz</strong> (ProjectName was extracted from the config.yml file). The index.html inside it is the bookdown report.</p>
</div>
<div id="ea2da" class="section level2" number="5.5">
<h2><span class="header-section-number">5.5</span> EA2DA</h2>
<pre><code>EA2DA A/path/to/a/config/file

#Example:
EA2DA ~/RNASequest/example/SRP199678/EA20220328_0/config.yml</code></pre>
<p>The execution of the above command will produce 6 data files which are required for the <a href="https://github.com/interactivereport/OmicsView" target="_blank">OmicsView</a> project import.</p>
<p><strong><em>Please fill in the empty entries in the Project_Info.csv before import.</em></strong></p>
</div>
<div id="administration" class="section level2" number="5.6">
<h2><span class="header-section-number">5.6</span> Administration</h2>
<p>There are two config files in the pipeline folder:</p>
<ul>
<li><p>config.tmp.yml: The template of the config file, with all default values;</p></li>
<li><p>sys.yml: the system config file, which includes:</p>
<ul>
<li><ol style="list-style-type: decimal">
<li>genome_path: the root path where the genome definition files (gtf) are located</li>
</ol></li>
<li><ol start="2" style="list-style-type: decimal">
<li>notCovariates: the column names from the sample meta information should not be considered as default covariates</li>
</ol></li>
<li><ol start="3" style="list-style-type: decimal">
<li>qc2meta: the column names from mapping QC file should be extracted and inserted into sample meta table</li>
</ol></li>
<li><ol start="4" style="list-style-type: decimal">
<li>QuickOmics_path: the file path to store the files for QuickOmics web server display</li>
</ol></li>
<li><ol start="5" style="list-style-type: decimal">
<li>DA_columns: the column names available for the sample meta table in the <a href="https://github.com/interactivereport/OmicsView" target="_blank">OmicsView</a> system</li>
</ol></li>
</ul></li>
</ul>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="data-preparation.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="output.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/interactivereport/RNASequest/edit/gh-pages/05-ExpressionAnalysisComponent.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/interactivereport/RNASequest/blob/gh-pages/05-ExpressionAnalysisComponent.Rmd",
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
