<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Installation | RNASequest: An end-to-end reproducible RNAseq data analysis and publishing framework</title>
  <meta name="description" content="Chapter 3 Installation | RNASequest: An end-to-end reproducible RNAseq data analysis and publishing framework" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Installation | RNASequest: An end-to-end reproducible RNAseq data analysis and publishing framework" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Installation | RNASequest: An end-to-end reproducible RNAseq data analysis and publishing framework" />
  
  
  



<meta name="date" content="2023-01-19" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="introduction.html"/>
<link rel="next" href="data-preparation.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">RNASequest</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Preface</a></li>
<li class="chapter" data-level="2" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>2</b> Introduction</a></li>
<li class="chapter" data-level="3" data-path="installation.html"><a href="installation.html"><i class="fa fa-check"></i><b>3</b> Installation</a>
<ul>
<li class="chapter" data-level="3.1" data-path="installation.html"><a href="installation.html#install-expressionanalysis"><i class="fa fa-check"></i><b>3.1</b> Install ExpressionAnalysis</a></li>
<li class="chapter" data-level="3.2" data-path="installation.html"><a href="installation.html#install-quickomics"><i class="fa fa-check"></i><b>3.2</b> Install Quickomics</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="installation.html"><a href="installation.html#install-r-packages"><i class="fa fa-check"></i><b>3.2.1</b> Install R packages</a></li>
<li class="chapter" data-level="3.2.2" data-path="installation.html"><a href="installation.html#clone-quickomics-github-repository"><i class="fa fa-check"></i><b>3.2.2</b> Clone Quickomics GitHub repository</a></li>
<li class="chapter" data-level="3.2.3" data-path="installation.html"><a href="installation.html#launch-the-r-shiny-app"><i class="fa fa-check"></i><b>3.2.3</b> Launch the R Shiny App</a></li>
<li class="chapter" data-level="3.2.4" data-path="installation.html"><a href="installation.html#configure-quickomics-to-load-projects-directly-from-urls"><i class="fa fa-check"></i><b>3.2.4</b> Configure QuickOmics to load projects directly from URLs</a></li>
<li class="chapter" data-level="3.2.5" data-path="installation.html"><a href="installation.html#prepare-project-data-to-be-loaded-from-url"><i class="fa fa-check"></i><b>3.2.5</b> Prepare project data to be loaded from URL</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-preparation.html"><a href="data-preparation.html"><i class="fa fa-check"></i><b>4</b> Data preparation</a>
<ul>
<li class="chapter" data-level="4.1" data-path="data-preparation.html"><a href="data-preparation.html#retrieving-data-from-ncbi-geo"><i class="fa fa-check"></i><b>4.1</b> Retrieving data from NCBI GEO</a></li>
<li class="chapter" data-level="4.2" data-path="data-preparation.html"><a href="data-preparation.html#file-hierarchy"><i class="fa fa-check"></i><b>4.2</b> File hierarchy</a></li>
<li class="chapter" data-level="4.3" data-path="data-preparation.html"><a href="data-preparation.html#self-prepared-files"><i class="fa fa-check"></i><b>4.3</b> Self-prepared files</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="expression-analysis-ea-component.html"><a href="expression-analysis-ea-component.html"><i class="fa fa-check"></i><b>5</b> Expression analysis (EA) component</a>
<ul>
<li class="chapter" data-level="5.1" data-path="expression-analysis-ea-component.html"><a href="expression-analysis-ea-component.html#eainit"><i class="fa fa-check"></i><b>5.1</b> EAinit</a></li>
<li class="chapter" data-level="5.2" data-path="expression-analysis-ea-component.html"><a href="expression-analysis-ea-component.html#eaqc"><i class="fa fa-check"></i><b>5.2</b> EAqc</a></li>
<li class="chapter" data-level="5.3" data-path="expression-analysis-ea-component.html"><a href="expression-analysis-ea-component.html#earun"><i class="fa fa-check"></i><b>5.3</b> EArun</a></li>
<li class="chapter" data-level="5.4" data-path="expression-analysis-ea-component.html"><a href="expression-analysis-ea-component.html#eareport"><i class="fa fa-check"></i><b>5.4</b> EAreport</a></li>
<li class="chapter" data-level="5.5" data-path="expression-analysis-ea-component.html"><a href="expression-analysis-ea-component.html#ea2da"><i class="fa fa-check"></i><b>5.5</b> EA2DA</a></li>
<li class="chapter" data-level="5.6" data-path="expression-analysis-ea-component.html"><a href="expression-analysis-ea-component.html#administration"><i class="fa fa-check"></i><b>5.6</b> Administration</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="output.html"><a href="output.html"><i class="fa fa-check"></i><b>6</b> Output</a>
<ul>
<li class="chapter" data-level="6.1" data-path="output.html"><a href="output.html#quickomics-visualization"><i class="fa fa-check"></i><b>6.1</b> QuickOmics visualization</a></li>
<li class="chapter" data-level="6.2" data-path="output.html"><a href="output.html#other-output-files"><i class="fa fa-check"></i><b>6.2</b> Other output files</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="advanced-plotting-functions.html"><a href="advanced-plotting-functions.html"><i class="fa fa-check"></i><b>7</b> Advanced plotting functions</a>
<ul>
<li class="chapter" data-level="7.1" data-path="advanced-plotting-functions.html"><a href="advanced-plotting-functions.html#pca-plot"><i class="fa fa-check"></i><b>7.1</b> PCA plot</a></li>
<li class="chapter" data-level="7.2" data-path="advanced-plotting-functions.html"><a href="advanced-plotting-functions.html#covariate-checking"><i class="fa fa-check"></i><b>7.2</b> Covariate checking</a></li>
<li class="chapter" data-level="7.3" data-path="advanced-plotting-functions.html"><a href="advanced-plotting-functions.html#volcano-plot"><i class="fa fa-check"></i><b>7.3</b> Volcano plot</a></li>
<li class="chapter" data-level="7.4" data-path="advanced-plotting-functions.html"><a href="advanced-plotting-functions.html#heatmap"><i class="fa fa-check"></i><b>7.4</b> Heatmap</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="quickomics-component.html"><a href="quickomics-component.html"><i class="fa fa-check"></i><b>8</b> Quickomics component</a></li>
<li class="chapter" data-level="9" data-path="bookdown-component.html"><a href="bookdown-component.html"><i class="fa fa-check"></i><b>9</b> Bookdown component</a></li>
<li class="chapter" data-level="10" data-path="slide-deck-component.html"><a href="slide-deck-component.html"><i class="fa fa-check"></i><b>10</b> Slide deck component</a></li>
<li class="chapter" data-level="11" data-path="shinyone-component.html"><a href="shinyone-component.html"><i class="fa fa-check"></i><b>11</b> ShinyOne component</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">RNASequest: An end-to-end reproducible RNAseq data analysis and publishing framework</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="installation" class="section level1" number="3">
<h1><span class="header-section-number">Chapter 3</span> Installation</h1>
<div id="install-expressionanalysis" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Install ExpressionAnalysis</h2>
<p>First we install the ExpressionAnalysis by downloading the scripts from GitHub:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="installation.html#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> clone https://github.com/interactivereport/RNASequest.git</span>
<span id="cb1-2"><a href="installation.html#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> RNASequest</span>
<span id="cb1-3"><a href="installation.html#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="installation.html#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Install RNASequest conda environment</span></span>
<span id="cb1-5"><a href="installation.html#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Please make sure you have conda installed before, and this step may take a while</span></span>
<span id="cb1-6"><a href="installation.html#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">bash</span> install</span>
<span id="cb1-7"><a href="installation.html#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="installation.html#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="co">#Activate the conda environment</span></span>
<span id="cb1-9"><a href="installation.html#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate ExpressionAnalysis</span>
<span id="cb1-10"><a href="installation.html#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="installation.html#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="co">#Also, .env will be created under the src directory:</span></span>
<span id="cb1-12"><a href="installation.html#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> ~/RNASequest/src/.env</span>
<span id="cb1-13"><a href="installation.html#cb1-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-14"><a href="installation.html#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="co">#Check the path of current directory and add it to $PATH:</span></span>
<span id="cb1-15"><a href="installation.html#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="va">CurrentDir</span><span class="op">=</span><span class="kw">`</span><span class="bu">pwd</span><span class="kw">`</span></span>
<span id="cb1-16"><a href="installation.html#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">PATH</span><span class="op">=</span><span class="st">&quot;</span><span class="va">$CurrentDir</span><span class="st">:</span><span class="va">$PATH</span><span class="st">&quot;</span></span>
<span id="cb1-17"><a href="installation.html#cb1-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-18"><a href="installation.html#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="co">#However, the above command only adds the RNASequest directory to $PATH temporarily</span></span>
<span id="cb1-19"><a href="installation.html#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="co">#To add it to the environment permanently, edit ~/.bash_profile or ~/.bashrc:</span></span>
<span id="cb1-20"><a href="installation.html#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="ex">vim</span> ~/.bash_profile</span>
<span id="cb1-21"><a href="installation.html#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="co">#Add the full path of the RNASequest directory to $PATH, for example, $HOME/RNASequest</span></span>
<span id="cb1-22"><a href="installation.html#cb1-22" aria-hidden="true" tabindex="-1"></a><span class="va">PATH</span><span class="op">=</span><span class="va">$PATH</span>:<span class="va">$HOME</span>/RNASequest</span>
<span id="cb1-23"><a href="installation.html#cb1-23" aria-hidden="true" tabindex="-1"></a><span class="co">#Source the file</span></span>
<span id="cb1-24"><a href="installation.html#cb1-24" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> ~/.bash_profile</span></code></pre></div>
<p>Set the default paths for ExpressionAnalysis:</p>
<p>In <a href="https://github.com/interactivereport/RNASequest/blob/main/src/sys_example.yml" target="_blank">sys_example.yml</a> file is used to set the default configurations of the whole pipeline, including paths for Quickomics visualization. Please modify this file before running the pipeline, since the paths will be different from server to server.</p>
<p><strong>Please use this sys_example.yml to create a sys.yml file under the src directory, and fill in the necessary information.</strong> Here is an example of the sys.yml file:</p>
<p>The QuickOmics_test_folder and QuickOmics_publish_folder are defined by the config.csv file in the QuickOmics app folder (See 3.2).</p>
<pre><code>## path to the reference genome is on HPC
genome_path: /path/to/genome/references
notCovariates: [Sapio_Submission_ID,Sapio_Sample_ID,Sapio_Request_Name,Sapio_Request_ID,Sapio_URL,Sapio_Plate_ID,Sapio_Profile,
                TST_ID,Status_Time,Well,Organism,ID,Sample_Process,Plate_Name,Volume, #Well_Row,
                Sample_Type,Sample_Name,Concentration, #Well_Column,
                Annotated_By,Ethnicity,Race]

#sample2meta: [Sapio_Plate_Name,RIN,Organ,Tissue,Cell_type,Cell_Line,Gender,Disease,Treatment,Timepoint,Genotype,Age]
qc2meta: [Intragenic Rate,Exonic Rate,Mapping Rate,Genes Detected,Mean Per Base Cov.,Estimated Library Size,Mapped,Intergenic Rate,Total Purity Filtered Reads Sequenced,Intronic Rate,Fragment Length Mean,rRNA rate,No. Covered 5&#39;,Duplication Rate of Mapped]
QuickOmics_test_folder: /path/to/QuickOmics_test_folder/
QuickOmics_test_link:  http://ngs.biogen.com:3838/Quickomics/?testfile=
QuickOmics_publish_folder: /path/to/QuickOmics_data/
QuiclOmics_publish_link: http://ngs.biogen.com:3838/Quickomics/?serverfile=
shinyApp: http://ngs.biogen.com/shinyone/app/core/
DA_columns: [SampleID, ProjectName, PlatformName, CellType, Collection, DataSource, Description, DiseaseStage, DiseaseState,
             Ethnicity, Gender, Infection, Organism, Response, SamplePathology, SampleSource, SampleType, 
             SamplingTime, SubjectID, Symptom, Tissue, Title, Transfection, Treatment, Age]
DNAnexus:
  prjID: FASTR_NGS_PROJECT_ID
  species: FASTR_REFERENCE_SPECIES
  ref: FASTR_REFERENCE_VERSION
  uName: FASTR_USER_FULLNAME
  count: [genes.estcount_table,genes.expected_count]
  effL: genes.effective_length
  tpm: [genes.tpm_table,genes.tpm]
  indFlag: genes.results.gz
  seqQC: [combined.metrics,merged_metrics]
FileName:
  projectFile: project.yml
  prj_counts: counts
  prj_effLength: length
  prj_seqQC: seqQC
  prj_TPM: tpm
  sample_meta: meta
  gene_annotation: annotation
  comparison_file: comparison
  projectEntry:
    prjID: project
    prjTitle: project
    species: species</code></pre>
</div>
<div id="install-quickomics" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> Install Quickomics</h2>
<div id="install-r-packages" class="section level3" number="3.2.1">
<h3><span class="header-section-number">3.2.1</span> Install R packages</h3>
<p><strong>Method 1.</strong> Create conda environment from yml file. This is the easiest way to install all packages. Please download the file <a href="https://github.com/interactivereport/Quickomics/blob/master/conda_environment/QuickOmics.yml" target="_blank">https://github.com/interactivereport/Quickomics/blob/master/conda_environment/QuickOmics.yml</a>, load your conda, then run:</p>
<pre><code>conda env create -f QuickOmics.yml
#The above command may take a while, then you will have a new conda environment called QuickOmics</code></pre>
<p><strong>Method 2.</strong> Install all packages in your current R environment. Depending on your system, you may need to install other required programs. We recommend using the conda environment method whenever possible.</p>
<pre><code>cran_packages=c(&quot;shiny&quot;, &quot;shinythemes&quot;, &quot;shinyjs&quot;, &quot;plotly&quot;, &quot;reshape2&quot;, &quot;tidyverse&quot;, &quot;gplots&quot;, &quot;ggpubr&quot;,
&quot;gridExtra&quot;, &quot;ggrepel&quot;, &quot;RcolorBrewer&quot;, &quot;pheatmap&quot;, &quot;rgl&quot;, &quot;car&quot;, &quot;colourpicker&quot;, &quot;VennDiagram&quot;, &quot;factoextra&quot;,
&quot;openxlsx&quot;, &quot;visNetwork&quot;, &quot;cowplot&quot;, &quot;circlize&quot;, &quot;svglite&quot;, &quot;shinyjqui&quot;, &quot;Hmisc&quot;, &quot;ggrastr&quot;,
&quot;ggExtra&quot;, &quot;networkD3&quot;, &quot;vctrs&quot;, &quot;ragg&quot;, &quot;textshaping&quot;, &quot;stringi&quot;, &quot;plyr&quot;, &quot;png&quot;, &quot;psych&quot;, &quot;broom&quot;)

#Note: Hmisc is not required to run the Shiny app but is needed to prepare network data from expression matrix.
install.packages(cran_packages, repos=&quot;http://cran.r-project.org/&quot;)  #choose repos based on your location if needed
if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;)
BiocManager::install(c(&quot;Mfuzz&quot;, &quot;biomaRt&quot;, &quot;ComplexHeatmap&quot;, &quot;pathview&quot;))

#Requirements: shiny &gt;= v1.4.0.2 </code></pre>
</div>
<div id="clone-quickomics-github-repository" class="section level3" number="3.2.2">
<h3><span class="header-section-number">3.2.2</span> Clone Quickomics GitHub repository</h3>
<pre><code>git clone https://github.com/interactivereport/Quickomics.git</code></pre>
<p>see more at <a href="https://docs.github.com/en/free-pro-team@latest/github/creating-cloning-and-archiving-repositories/cloning-a-repository" target="_blank">https://docs.github.com/en/free-pro-team@latest/github/creating-cloning-and-archiving-repositories/cloning-a-repository</a></p>
</div>
<div id="launch-the-r-shiny-app" class="section level3" number="3.2.3">
<h3><span class="header-section-number">3.2.3</span> Launch the R Shiny App</h3>
<p>Check the following web links on various options to launch the app.</p>
<p><a href="https://shiny.rstudio.com/articles/running.html" target="_blank">https://shiny.rstudio.com/articles/running.html</a></p>
<p><a href="https://shiny.rstudio.com/deploy/" target="_blank">https://shiny.rstudio.com/deploy/</a></p>
</div>
<div id="configure-quickomics-to-load-projects-directly-from-urls" class="section level3" number="3.2.4">
<h3><span class="header-section-number">3.2.4</span> Configure QuickOmics to load projects directly from URLs</h3>
<p>QuickOmics has a feature to load a project directly from URL in the form of <a href="https://quickomics.bxgenomics.com/?serverfile=SRP199678" target="_blank">https://quickomics.bxgenomics.com/?serverfile=SRP199678</a>, which requires saving project data files to a user-specified directory on the server hosting QuickOmics. To use this feature, create a file called config.csv in the QuickOmics App folder.</p>
<pre><code>category,value
server_dir,{Path/to/Server/File}
test_dir,{Path/to/Test/Files}</code></pre>
<p>Files stored in server_dir can be loaded by URLs like: <a href="https://quickomics.bxgenomics.com/?serverfile=project_ID" target="_blank">https://quickomics.bxgenomics.com/?serverfile=project_ID</a></p>
<p>Files stored in test_dir can be loaded by URLs like: <a href="https://quickomics.bxgenomics.com/?testfile=project_ID" target="_blank">https://quickomics.bxgenomics.com/?testfile=project_ID</a></p>
</div>
<div id="prepare-project-data-to-be-loaded-from-url" class="section level3" number="3.2.5">
<h3><span class="header-section-number">3.2.5</span> Prepare project data to be loaded from URL</h3>
<p>To prepare project files to be loaded via URL, do the following.</p>
<p>For each project, besides the two RData files (projectID.RData and ProjectID_netword.RData), prepare a ProjectID.csv file with six columns, e.g.</p>
<pre><code>&quot;Name&quot;,&quot;ShortName&quot;,&quot;ProjectID&quot;,&quot;Species&quot;,&quot;ExpressionUnit&quot;,&quot;Path&quot;
&quot;RNAseq analysis of sorted microglia&quot;,&quot;SRP199678&quot;,&quot;SRP199678&quot;,&quot;mouse&quot;,&quot;log2(TPM+0.25)&quot;,&quot;/camhpc/home/ysun4/RNASequest/example/SRP199678/EA20220329_0&quot;</code></pre>
<p>The last two columns in the csv file are optional but recommended.</p>
<p>Now copy the three files (projectID.RData, ProjectID_netword.RData and ProjectID.csv) inside the server_dir folder as specified by the config.csv file, and the project can be loaded as <a href="https://quickomics.bxgenomics.com/?serverfile=project_ID" target="_blank">https://quickomics.bxgenomics.com/?serverfile=project_ID</a> (replace quickomics.bxgenomics.com by the URL of your own QuickOmics instance).</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="data-preparation.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/interactivereport/RNASequest/edit/gh-pages/03-Installation.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/interactivereport/RNASequest/blob/gh-pages/03-Installation.Rmd",
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
