<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 Running EA pipeline | RNASequest: An end-to-end reproducible RNAseq data analysis and publishing framework</title>
  <meta name="description" content="Chapter 5 Running EA pipeline | RNASequest: An end-to-end reproducible RNAseq data analysis and publishing framework" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 Running EA pipeline | RNASequest: An end-to-end reproducible RNAseq data analysis and publishing framework" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 Running EA pipeline | RNASequest: An end-to-end reproducible RNAseq data analysis and publishing framework" />
  
  
  



<meta name="date" content="2022-04-04" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="data-preparation.html"/>
<link rel="next" href="output.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">RNASequest</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Preface</a></li>
<li class="chapter" data-level="2" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>2</b> Introduction</a></li>
<li class="chapter" data-level="3" data-path="installation.html"><a href="installation.html"><i class="fa fa-check"></i><b>3</b> Installation</a></li>
<li class="chapter" data-level="4" data-path="data-preparation.html"><a href="data-preparation.html"><i class="fa fa-check"></i><b>4</b> Data preparation</a>
<ul>
<li class="chapter" data-level="4.1" data-path="data-preparation.html"><a href="data-preparation.html#retrieving-data-from-ncbi-geo"><i class="fa fa-check"></i><b>4.1</b> Retrieving data from NCBI GEO</a></li>
<li class="chapter" data-level="4.2" data-path="data-preparation.html"><a href="data-preparation.html#file-hierarchy"><i class="fa fa-check"></i><b>4.2</b> File hierarchy</a></li>
<li class="chapter" data-level="4.3" data-path="data-preparation.html"><a href="data-preparation.html#self-prepared-files"><i class="fa fa-check"></i><b>4.3</b> Self-prepared files</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="running-ea-pipeline.html"><a href="running-ea-pipeline.html"><i class="fa fa-check"></i><b>5</b> Running EA pipeline</a>
<ul>
<li class="chapter" data-level="5.1" data-path="running-ea-pipeline.html"><a href="running-ea-pipeline.html#eainit"><i class="fa fa-check"></i><b>5.1</b> EAinit</a></li>
<li class="chapter" data-level="5.2" data-path="running-ea-pipeline.html"><a href="running-ea-pipeline.html#eaqc"><i class="fa fa-check"></i><b>5.2</b> EAqc</a></li>
<li class="chapter" data-level="5.3" data-path="running-ea-pipeline.html"><a href="running-ea-pipeline.html#earun"><i class="fa fa-check"></i><b>5.3</b> EArun</a></li>
<li class="chapter" data-level="5.4" data-path="running-ea-pipeline.html"><a href="running-ea-pipeline.html#eareport"><i class="fa fa-check"></i><b>5.4</b> EAreport</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="output.html"><a href="output.html"><i class="fa fa-check"></i><b>6</b> Output</a>
<ul>
<li class="chapter" data-level="6.1" data-path="output.html"><a href="output.html#quickomics-visualization"><i class="fa fa-check"></i><b>6.1</b> QuickOmics visualization</a></li>
<li class="chapter" data-level="6.2" data-path="output.html"><a href="output.html#other-output-files"><i class="fa fa-check"></i><b>6.2</b> Other output files</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="bookdown-report.html"><a href="bookdown-report.html"><i class="fa fa-check"></i><b>7</b> Bookdown report</a></li>
<li class="chapter" data-level="8" data-path="online-slide-deck.html"><a href="online-slide-deck.html"><i class="fa fa-check"></i><b>8</b> Online slide deck</a></li>
<li class="chapter" data-level="9" data-path="shinyone.html"><a href="shinyone.html"><i class="fa fa-check"></i><b>9</b> ShinyOne</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">RNASequest: An end-to-end reproducible RNAseq data analysis and publishing framework</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="running-ea-pipeline" class="section level1" number="5">
<h1><span class="header-section-number">Chapter 5</span> Running EA pipeline</h1>
<p>First, activate the conda environment if you havenâ€™t done so:</p>
<pre><code>conda activate ExpressionAnalysis</code></pre>
<div id="eainit" class="section level2" number="5.1">
<h2><span class="header-section-number">5.1</span> EAinit</h2>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="running-ea-pipeline.html#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">EAinit</span> Path/to/a/DNAnexus/result/folder</span>
<span id="cb11-2"><a href="running-ea-pipeline.html#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="running-ea-pipeline.html#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Example:</span></span>
<span id="cb11-4"><a href="running-ea-pipeline.html#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="ex">EAinit</span> ~/RNASequest/example/SRP199678</span></code></pre></div>
<p>The execution of above command will create a sub-folder (EA[timestamp]) in the specified RNAseq result folder.
There will be five files in the result folder:</p>
<ul>
<li>compareInfo.csv: an empty comparison definition file (with header). Please fill in this file before <code>EArun</code> call.</li>
<li>config.yml: a config fill specifies the parameters of the <code>EAqc</code> and <code>EArun</code>. Please update <strong>covariates_adjust</strong> after <code>EAqc</code>.</li>
<li>geneAnnotation.csv: a gene annotation file including gene symbol.</li>
<li>sampleMeta.csv: a sample meta information file, please feel free to add additional columns whose column names should be considered to be added into <strong>covariates_check</strong> in <em>config.yml</em>.</li>
<li>alignQC.pdf: plots generated from alignment QC metrics.</li>
</ul>
<p><strong><em>Please pay attention on the std out messages.</em></strong></p>
</div>
<div id="eaqc" class="section level2" number="5.2">
<h2><span class="header-section-number">5.2</span> EAqc</h2>
<pre><code>EAqc Path/to/a/config/file

#Example:
EAqc ~/RNASequest/example/SRP199678/EA20220328_0/config.yml</code></pre>
<p>The execution of the command with the above default config file, expression PC analysis will be done against covariates specified in <strong>covariates_check</strong> in <em>config.yml</em> file. An excel file will list p-value for all numeric and categorical covariates, and with significant ones will be in plot pdf files. The analysis before covariate adjusting will have prefix <em>covariatePCanalysis_noAdjust</em>.
Based on the above results, you can add covariates into <strong>covariates_adjust</strong> in <em>config.yml</em> file, and run <code>EAqc</code> again. This time additional expression PC analysis will be applied to covariate adjusted expression with files started with <em>covariatePCanalysis_Adjusted</em>.</p>
<p><strong><em>Please pay attention on the std out messages.</em></strong></p>
</div>
<div id="earun" class="section level2" number="5.3">
<h2><span class="header-section-number">5.3</span> EArun</h2>
<pre><code>EArun Path/to/a/config/file

#Example:
EArun ~/RNASequest/example/SRP199678/EA20220328_0/config.yml</code></pre>
<p><strong><em>Please fill the compareInfo.csv before executing the above command</em></strong></p>
<p>The execution of above command will produce R objects for QuickOmics webserver to load. The process will generate the covariate adjusted logTPM for visualization; complete differentially expressed gene analysis and gene network generation.</p>
<p>The results (four files) can be uploaded to the QuickOmics webserver.</p>
<p><strong><em>Please pay attention on the std out messages.</em></strong></p>
</div>
<div id="eareport" class="section level2" number="5.4">
<h2><span class="header-section-number">5.4</span> EAreport</h2>
<pre><code>EAreport Path/to/a/config/file

#Example:
EAreport ~/RNASequest/example/SRP199678/EA20220328_0/config.yml</code></pre>
<p>By running the command above, the pipeline will generate a <strong>BookdownReport</strong> folder in the same directory as the config file. This folder contains the raw Rmd files, as well as the final bookdown report, which is the BookdownReport/docs/index.html file. If you move the directory, please move everything inside the docs directory, since other files are part of the bookdown report, and the index.html can automatically link everything together.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="data-preparation.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="output.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/interactivereport/RNASequest/edit/gh-pages/05-RunningPipeline.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/interactivereport/RNASequest/blob/gh-pages/05-RunningPipeline.Rmd",
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
